### PR TITLE
Change log server URL locally

### PR DESCRIPTION
**Pull request by avazirna**
## Summary
This PR adds a new option to the `Developer Options` menu that allows users the set a new URL for where CommCare logs should be submitted.

Ticket: https://dimagi.atlassian.net/browse/SAAS-15980

## Product Description
Users are able to set the URL by typing it in the text box and remove it by leaving the text box blank:
<video width="200" src="https://github.com/user-attachments/assets/f167c52b-ed63-487f-a743-48d115cd9d0d"/>

## Safety Assurance

- [X] I have confidence that this PR will not introduce a regression for the reasons below
- [X] Do we need to enhance manual QA test coverage ? If yes, "QA Note" label is set correctly

### Safety story
Local tests were successful but this should also be tested by QA.

### CODE REVIEW

File: app/res/xml/preferences_developer.xml
Comment (shubham1g5) [inline app/res/xml/preferences_developer.xml:177]: Wondering if user ever sees this `defaultValue` as an app should always have the `log_receiver_url` populated ?

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (shubham1g5) [inline app/src/org/commcare/preferences/DeveloperPreferences.java:131]: think we should be defaulting to the url from app profile xml instead of completely removing it.

**Review by shubham1g5 — CHANGES_REQUESTED** [review]
Few points/questions - 

- Can we add a link to the ticket or github issue which is prompting the change
-  Are we clear and ok with set of steps one need to take to see this setting in dev options ? (Dev options visibility defer in debug and release builds)
- Have we confirmed with the client that this solution works with them ?
- Is it urgent enough for the client to include in a hotfix ?

File: app/res/xml/preferences_developer.xml
Comment (avazirna) [reply app/res/xml/preferences_developer.xml:177]: they see it when `log_receiver_url` is not set, by [default the logs URL is the Form Submission log](https://github.com/dimagi/commcare-android/blob/831747a00d29328ea0aaa886245e86588fc4af84/app/src/org/commcare/tasks/LogSubmissionTask.java#L112)

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (avazirna) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:131]: I think this is related to the notion of what is the default URL for the logs submission, right now it seems to be the form submission URL

**Comment by avazirna** [comment]
> Few points/questions -
> 
> * Can we add a link to the ticket or github issue which is prompting the change

Done
> * Are we clear and ok with set of steps one need to take to see this setting in dev options ? (Dev options visibility defer in debug and release builds)

Good point, I may have to whitelist this option, let me check.
> * Have we confirmed with the client that this solution works with them ?

I got sign-off from [this](https://dimagi.atlassian.net/browse/SAAS-15980?focusedCommentId=347077) comment about this option. And I think we are still waiting for the client's response about it [here](https://dimagi.atlassian.net/browse/SUPPORT-21952?focusedCommentId=347123).
> * Is it urgent enough for the client to include in a hotfix ?

Given the context around this client, I believe this is the bare minimum to satisfy their request and considering that they are in the middle of investigating the issues, it does seem relevant enough to take this opportunity to roll it out now.

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (shubham1g5) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:131]: yeah we should be defaulting it to form submission url instead of removing it completely.

**Comment by ctsims** [comment]
@avazirna I think it's worth confirming briefly that this option can cleanly be set to a different URL, used, then have the custom text deleted, and make sure that the submission reverts to the original receiver. With settings like this where the option is a String value or an empty value, there's always a risk that a blank string setting is treated differently than a null or empty value.

**Comment by avazirna** [comment]
> @avazirna I think it's worth confirming briefly that this option can cleanly be set to a different URL, used, then have the custom text deleted, and make sure that the submission reverts to the original receiver. With settings like this where the option is a String value or an empty value, there's always a risk that a blank string setting is treated differently than a null or empty value.

Yes @ctsims, I tested that scenario a couple of times and also added a [listener to remove the preference](https://github.com/dimagi/commcare-android/pull/2852/commits/a2be44ba7aa194f4692c0364914774d5a70b88ed) when `blank`

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (avazirna) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:132]: @shubham1g5 that would be changing the way this preference is currently used and create redundancy. Right now, it seems that we only set the preference when the logs URL is different from the form submissions URL. Defaulting to the submissions URL and to keep consistency, would require setting it during app installation as well, right?

Comment (ctsims) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:132]: @shubham1g5 that was my first impulse as well, but I think it's actually too dangerous to set this redundantly for two reasons.

1 - It introduces a new parity requirement. If we changed the submission URL for form submissions on HQ's end, we'd now need to **also** change the log URL, but we can't really do that with HQ settings because it'll blow away the config for anyone who's added the custom setting.
2 - From a usability perspective, it's dangerous because the users don't have a super meaningful way currently to know what **was** in the box for this current UI before. If the Log URL is Empty when you load it, it's clear that setting it back to blank should be the "inverse" that sets things back to neutral. If you load up the field then and its got the HQ default form URL already in it, I think you'd naturally then assume that you need to keep track of that URL so you can set it back if you want to undo the action.

**Comment by avazirna** [comment]
@damagatchi retest this please

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (shubham1g5) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:132]: > Defaulting to the submissions URL and to keep consistency, would require setting it during app installation as well, right?

Makes sense, I didn't realise we don't do it already. 

> we can't really do that with HQ settings because it'll blow away the config for anyone who's added the custom setting.

I am not sure I understand it. My thought was to do this as a one time action at app install time -

1. While Installing app, we populate the `logs_receiver_url` in app preferences by copying it from form submission url.

2. When user opens the logs url preference to edit, they would see what this url is pointing to (which would be same as form submission url)

3. User edits the url to some other valid url. We preserve it and show it to user in this preference

4. User edits the url and change is to a non-valid url (empty or something else) -> We show user an error that the change they made is invalid and preserve whatever the old value of the url was. 


No strong thoughts here though overall, I think we want to preserve the existing behaviour in the PR, probably better to rename the field from `Log Submission Server` to `Custom Log Submission Server` to make it more clear that existing log server is not empty when they see it empty.

**Comment by avazirna** [comment]
@damagatchi retest this please

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (avazirna) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:132]: @shubham1g5 I think it's a matter of interpretation, my understanding from going through the code is that we don't have a `Log Submission Server` per se, everything is to be sent to the `Form Submission Server`. However, we can have the option to set a `Log Submission Server`. Is this a valid interpretation? I'm happy to follow the `Custom Log Submission Server` approach

File: app/res/values/strings.xml
Comment (ctsims) [inline app/res/values/strings.xml:459]: I think this is going to be really confusing if this is the default value. 

From my perspective and the discussion below, we need to frame this as an "override" setting, which redirects logs to a separate path. Having any value in here (like a stub URL) feels very confusing, as it implies that the value was in use before.

I think the "baseline" value of this setting should be an empty string (``) so it's clear that its clear when it's unused, but more critically so that it's clearer on how to reset back to empty.

File: app/res/xml/preferences_developer.xml
Comment (ctsims) [inline app/res/xml/preferences_developer.xml:180]: As per the discussion  below, I think we should rename the text on this to be something like "Override: Custom Log Submission Server" so that it's clear that this is a secondary setting, not the baseline.

File: app/src/org/commcare/preferences/DeveloperPreferences.java
Comment (ctsims) [reply app/src/org/commcare/preferences/DeveloperPreferences.java:132]: Agreed with "Custom" and updated suggestions on how to be consistent with my take on this above. I think it'd introduce a lot of ambiguity to create a new situation where we **expect** the log submission URL to be set, rather than treating this as an override. 

@shubham1g5 I think the big drawback in your user story is that currently users have an expectation that this URL is synchronized, and the 'custom' version is a rare edge case, so I'd be concerned about situations where we'd need to keep the two in sync across other events or changes by default.

File: app/res/values/strings.xml
Comment (avazirna) [reply app/res/values/strings.xml:459]: I was trying to find a way to inform the user that the protocol needs to be `https` as it will fail otherwise. But happy to explore another option

**Comment by avazirna** [comment]
@damagatchi retest this please

File: app/res/xml/preferences_developer.xml
Comment (avazirna) [reply app/res/xml/preferences_developer.xml]: Done 9e623dfe3579c385de1f1efe58a49dc5cbb95cb7

**Comment by avazirna** [comment]
@damagatchi retest this please

File: app/res/xml/preferences_developer.xml
Comment (shubham1g5) [inline app/res/xml/preferences_developer.xml:177]: nit: We should also rename it to have the same `Override` as title.

**Review by shubham1g5 — APPROVED** [review]

File: app/res/xml/preferences_developer.xml
Comment (avazirna) [reply app/res/xml/preferences_developer.xml]: And I forgot the word "Custom" too. Corrected [0069632](https://github.com/dimagi/commcare-android/pull/2852/commits/0069632157b68b90c94cf59b3d6d5acf4d9e34d6)

Comment (shubham1g5) [reply app/res/xml/preferences_developer.xml]: Ahh I think if we use `Custom` , we should remove the `Override` i.e. use one of them instead of both.

Comment (avazirna) [reply app/res/xml/preferences_developer.xml]: Agree

**Pull request merged**
